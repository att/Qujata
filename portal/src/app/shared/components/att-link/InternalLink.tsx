import { useMemo } from 'react';
import { Link as RouterLink, Path, useResolvedPath } from 'react-router-dom';
import { BaseLinkProps } from './Link.model';
import { getLinkClassName } from './Link.util';

export interface InternalLinkProps extends BaseLinkProps {
  queryParams?: string;
  replaceUrl?: boolean;
}

export const InternalLink: React.FC<InternalLinkProps> = (props: InternalLinkProps) => {
  const pathname: string = useFixReactRouterRelativePathname(props.link);
  return (
    <RouterLink
      to={{
        pathname,
        search: props.queryParams,
      }}
      replace={props.replaceUrl}
      className={getLinkClassName(props)}
      onClick={props.onLinkClick}
      title={props.linkTitle}
      data-cy={props.e2eId}
    >
      {props.children}
    </RouterLink>
  );
};

/*
  https://reactrouter.com/docs/en/v6/upgrading/v5#note-on-link-to-values
  react router Link component support link back "up" to parent routes, using a leading '..' segment
  From some reason it's working fine when "to" is a string but if it's an object it's not working
  To solve it, we will resolve the path here and send the absolute  path to Link component
*/
function useFixReactRouterRelativePathname(pathname: string): string {
  const memoizedTo: Partial<Path> = useMemo<Partial<Path>>(() => ({ pathname }), [pathname]);
  return useResolvedPath(memoizedTo).pathname;
}
