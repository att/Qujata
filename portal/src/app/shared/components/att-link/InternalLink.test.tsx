import { render, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { InternalLink, InternalLinkProps } from './InternalLink';
import { LinkSize, LinkStyle } from './Link.model';

describe('InternalLink', () => {
  test('renders large primary link', () => {
    const props: InternalLinkProps = {
      link: 'test-link',
      styleType: LinkStyle.PRIMARY,
      size: LinkSize.LARGE,
      children: undefined,
    };
    const { container }: RenderResult = render(<InternalLink {...props}>large primary link</InternalLink>, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders medium secondary link', () => {
    const props: InternalLinkProps = {
      link: 'test-link',
      styleType: LinkStyle.SECONDARY,
      size: LinkSize.MEDIUM,
      children: undefined,
    };
    const { container }: RenderResult = render(<InternalLink {...props}>medium secondary link</InternalLink>, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders regular link', () => {
    const props: InternalLinkProps = {
      link: 'test-link',
      styleType: LinkStyle.NONE,
      size: LinkSize.NONE,
      queryParams: 'test=2',
      children: undefined,
    };
    const { container }: RenderResult = render(<InternalLink {...props}>regular link</InternalLink>, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders wrapper link', () => {
    const props: InternalLinkProps = {
      link: 'test-link',
      styleType: LinkStyle.WRAPPER,
      size: LinkSize.NONE,
      replaceUrl: true,
      className: 'test-class-name',
      children: undefined,
    };
    const { container }: RenderResult = render(<InternalLink {...props}><span>wrapper link</span></InternalLink>, { wrapper: MemoryRouter });
    expect(container.firstChild).toMatchSnapshot();
    expect(container.firstChild).toHaveClass('test-class-name');
  });
});
