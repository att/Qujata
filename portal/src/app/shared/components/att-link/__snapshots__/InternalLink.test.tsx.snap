// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalLink renders large primary link 1`] = `
<a
  class="lg_link primary_link"
  href="/test-link"
>
  large primary link
</a>
`;

exports[`InternalLink renders medium secondary link 1`] = `
<a
  class="md_link secondary_link"
  href="/test-link"
>
  medium secondary link
</a>
`;

exports[`InternalLink renders regular link 1`] = `
<a
  class=""
  href="/test-link?test=2"
>
  regular link
</a>
`;

exports[`InternalLink renders wrapper link 1`] = `
<a
  class="test-class-name wrapper_link"
  href="/test-link"
>
  <span>
    wrapper link
  </span>
</a>
`;
